tests/framework/results/test_context.py

```diff
@@ -443,3 +445,130 @@ def test__bad_aggregator_stratification():
+        population=BASE_POPULATION, pop_filter=pop_filter
+    )
+    if pop_filter:
+        # The pop filter is set up to filter out everyone
+        assert filtered_pop.empty
+    else:
+        # An empty pop filter should return the entire population
+        assert filtered_pop.equals(BASE_POPULATION)
+
+
+@pytest.mark.parametrize("stratifications", [("familiar",), ()])
+def test__get_groups(stratifications):
+    groups = ResultsContext()._get_groups(
+        stratifications=stratifications, filtered_pop=BASE_POPULATION
+    )
+    assert isinstance(groups, DataFrameGroupBy)
+    if stratifications:
+        # Check that all familiars exist
+        assert set(groups.groups.keys()) == set(FAMILIARS)
+        # Check that the entire population is included
+        assert sum([len(value) for value in groups.groups.values()]) == len(BASE_POPULATION)
+    else:
+        item = groups.groups.popitem()
+        # Check that there are no other groups
+        assert not groups.groups
+        # Check that the group is 'all' and includes the entire population
+        key, val = item
+        assert key == "all"
+        assert val.equals(BASE_POPULATION.index)
+
+
+@pytest.mark.parametrize(
+    "stratifications, aggregator_sources",
+    [
+        (("familiar",), ["power_level"]),
+        (("familiar",), []),
+        (("familiar", "house"), ["power_level"]),
+        (("familiar", "house"), []),
+        ((), ["power_level"]),
+        ((), []),
+    ],
+)
+def test__aggregate(stratifications, aggregator_sources):
+    """Test that we are aggregating correctly. There are some nuances here:
+      - If aggregator_resources is provided, then simply .apply it to the groups passed in.
+      - If no aggregator_resources are provided, then we want a full aggregatio of the groups.
+
+    Note that the groups can be either a pandas DataFrame or a DataFrameGroupBy.
+    """
+    groups = ResultsContext()._get_groups(
+        stratifications=stratifications, filtered_pop=BASE_POPULATION
+    )
+    aggregates = ResultsContext()._aggregate(
+        pop_groups=groups,
+        aggregator_sources=aggregator_sources,
+        aggregator=len,
+    )
+    if stratifications:
+        full_idx = (
+            set(itertools.product(*(FAMILIARS, CATEGORIES)))
+            if "house" in stratifications
+            else set(FAMILIARS)
+        )
+        assert set(aggregates.index) == full_idx
+        group_lengths = BASE_POPULATION.groupby(list(stratifications)).apply(len)
+        assert (aggregates == group_lengths).all()
+    else:
+        assert len(aggregates.values) == 1
+        assert aggregates.values[0] == len(BASE_POPULATION)
+
+
+@pytest.mark.parametrize(
+    "aggregates",
+    [
+        pd.Series(data=[1, 2, 3], index=pd.Index(["a", "b", "c"], name="index")),
+        pd.DataFrame(data=[[1, 2, 3], [4, 5, 6]], index=pd.Index(["a", "b"], name="index")),
+    ],
+)
+def test__coerce_to_dataframe(aggregates):
```

Comment:
Add failure case test, for when the aggregator returns a dataframe with multiple columns (or not a dataframe at all).